Remote Desktop Protocol (RDP) matcher

vnxme opened the pull request:
The PR title seems to be self-explanatory. With this `rdp` matcher layer4 module allows Caddy to multiplex RDP connections. What is more, its `cookie_hash` option allows to match `mstshash` optional cookie value, and `cookie_ips` and `cookie_ports` options provide for `msts` optional cookie value (see important caveats in the comments to the matcher code). This way, if Microsoft Terminal Services Client (mstsc.exe) is used, Caddy may also route RDP connections to different serves based on the username, which is truncated to 9 characters to become `mstshash`. In addition, this matcher ads `cookie_hash`, `cookie_ip`, `cookie_port` and `correlation_id` to the replacer, if provided in the first packet. Last but not least, caddyfile support is included.

Simple RDP multiplexing
```json
{
  "apps": {
    "layer4": {
      "servers": {
        "example": {
          "listen": ["0.0.0.0:443"],
          "routes": [
            {
              "handle": [
                {
                  "handler": "proxy",
                  "upstreams": [
                    {"dial": ["single.machine.local:3389"]}
                  ]
                }
              ],
              "match": [
                {"rdp": {}}
              ]
            },
            {
              "handle": [
                {
                  "handler": "tls"
                },
                {
                  "handler": "proxy",
                  "upstreams": [
                    {"dial": ["localhost:8080"]}
                  ]
                }
              ],
              "match": [
                {"tls": {}}
              ]
            }
          ]
        }
      }
    }
  }
}
```

Same as above, but route RDP to 2 machines depending on cookie_hash, with a fallback mechanism
```json
{
  "apps": {
    "layer4": {
      "servers": {
        "example": {
          "listen": ["0.0.0.0:443"],
          "routes": [
            {
              "handle": [
                {
                  "handler": "proxy",
                  "upstreams": [
                    {"dial": ["another.machine.local:3389"]}
                  ]
                }
              ],
              "match": [
                {"rdp": {"cookie_hash":"Administr"}}
              ]
            }
            {
              "handle": [
                {
                  "handler": "proxy",
                  "upstreams": [
                    {"dial": ["fallback.machine.local:3389"]}
                  ]
                }
              ],
              "match": [
                {"rdp": {}}
              ]
            },
            {
              "handle": [
                {
                  "handler": "tls"
                },
                {
                  "handler": "proxy",
                  "upstreams": [
                    {"dial": ["localhost:8080"]}
                  ]
                }
              ],
              "match": [
                {"tls": {}}
              ]
            }
          ]
        }
      }
    }
  }
}
```

Tested on my machines having Windows 11 with a native server/client and MacOS Sonoma 14.5 with Jump Desktop 8.10.4 and Microsoft Remote Desktop 10.9.6 as clients. Please share your test results and capture the first packet with Wireshark, if the matcher doesn't work for you.

mholt reviewed (approved):
This is awesome!

I don't have windows to test this but I see the tests and I'm happy about those.

If it works for you I'm happy to accept this and we can iterate on it later if needed.

Very cool!

vnxme commented:
Rebased this PR onto master with #218 merged.

vnxme commented:
Wow, that was quick!

I added a few nice features, like custom info parsing and regexp matching, to better address RDP client diversity issues, and updated the opening post above.
